design/004-IoT-Adoption-Watcher.md

```diff
@@ -0,0 +1,248 @@
+
+### Interaction between the watcher and the main component
+
+Interaction between the watcher and main component based using the above
+interfaces is shown in the diagram below.
+
+![Interaction between the main component and watcher](004/watcher.svg)
+
+Figure 1: Interaction between the main component and watcher.
+
+### How different types of channels should be handled by watcher
+
+#### When main component registers a channel with the watcher
+
+1. For ledger channel: only its channel ID and latest off-chain state at the
+   time of registering is sufficient.
+
+   After registering a ledger channel, it is the main component's responsibility
+   to individually register all its sub-channels and virtual channels.
+
+2. For sub-channel: sub-channel ID, parent ledger channel ID and, the latest
+   off-chain states of both the sub-channel and parent ledger channel are
+   required.
+
+3. For virtual channel: virtual channel ID, relevant parent channel ID and, the
+   latest off-chain states of both the virtual channel and the relevant parent
+   ledger channel are required.
+
+   Relevant parent ledger channel is the ledger channel between this user and
+   the common intermediary.
+
+#### When main component sends new off-chain states for the watcher
+
+For all types of channels, watcher will have to store the newer state and
+discard the older ones.
+
+#### When registering states on the blockchain
+
+1. For ledger channel: the latest off-chain states of the ledger channel and,
+   of its sub-channels, virtual channels must be registered.
+
+2. In case sub-channel; the latest off-chain state of the parent ledger
+   channel, of the sub-channel and, of all the other sub-channels, virtual
+   channels funded by the parent ledger channel must be registered.
+
+2. In case virtual channel; the latest off-chain state of the relevant parent
+   ledger channel, of the sub-channel and, of all the other sub-channels,
+   virtual channels funded by the relevant parent ledger channel must be
+   registered.
+
+   Relevant parent ledger channel is the ledger channel between this user and
+   the common intermediary.
+
+## Rationale
+
+<!-- Provide a discussion of alternative approaches and trade offs; advantages
+and disadvantages of the specified approach.  -->
+
+The proposed design
+
+- Ensures watching component can be run either locally or as a remote service
+  and be connected to the blockchain in either case.
+
+- Ensures a channel instance can register with one or more watching components.
+  This is achieved by keeping the control of initializing all the pub-sub
+  components with the channel instance.
+
+## Impact
+
+<!-- Choose the level of impact this proposal will have: -->
+
+<!-- Minor (Does not impact any existing features) -->
+<!-- Major (Breaks one or more existing features) -->
+<!-- New Feature (Introduces a functionality) -->
+Architecture (Requires a modification of the architecture)
+
+## Implementation
```

Comment:
I am missing details on what needs to be changed in go-perun to integrate the watcher concept.

Reply:
I have briefly described the changes required in go-perun in points 1 and 2 of implementation. These are the only two changes, I foresee now. After these changes, I think the watcher could be implemented. 

Do you think, we should add a more detailed description of changes required ?

After the watcher is implemented, it would be upto the user of go-perun framework to initialize the watcher and register channels to it. On the other hand, we could also provide an option to make go-perun internally start and manage the watcher. What do you think ?